BootloaderCommonPkg/Library/TimeStampLib/TimeStampLib.c

```diff
@@ -44,3 +45,85 @@ GetTimeStampFrequency (
   // Ratio * 100000
   return (UINT32)(Ratio * 100000);
 }
+
+/**
+  Get timestamp accurate frequency in HZ by CPUID.
+
+  The TSC counting frequency is determined by using CPUID leaf 0x15. Frequency in MHz = Core XTAL frequency * EBX/EAX.
+  In newer flavors of the CPU, core xtal frequency is returned in ECX or 0 if not supported.
+
+  @retval The number of TSC counts per second.
+
+**/
+UINT64
+EFIAPI
+GetTimeStampAccurateFrequency (
+  VOID
+  )
+{
+  UINT64                 TscFrequency;
+  UINT32                 RegEax;
+  UINT32                 RegEbx;
+  UINT32                 RegEcx;
+
+  // Use CPUID leaf 0x15 Time Stamp Counter and Nominal Core Crystal Clock Information
+  // EBX returns 0 if not supported. ECX, if non zero, provides Core Xtal Frequency in hertz.
+  // TSC frequency = (ECX, Core Xtal Frequency) * EBX/EAX.
+  AsmCpuid (CPUID_TIME_STAMP_COUNTER, &RegEax, &RegEbx, &RegEcx, NULL);
+
+  // If EAX or EBX returns 0, the XTAL ratio is not enumerated.
+  if ((RegEax == 0) || (RegEbx == 0 )) {

```

Comment:
According to above comment, ECX could be also 0. The case may need to fallback to GetTimeStampFrequency as well.

Reply:
Good point, will make the change.

Reply:
Fixed it in new commit.